### PR TITLE
Creation of OAuth route for MK

### PR DESCRIPTION
**Pull request by SidharthBansal**
Fixes https://github.com/publiclab/mapknitter/issues/378

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@
 class SessionsController < ApplicationController
   #protect_from_forgery :except => [:create]
 
-  @@openid_url_base  = "https://publiclab.org/people/"
+  @@openid_url_base  = "localhost:3000/people/"
```
Comment (SidharthBansal) [inline app/controllers/sessions_controller.rb]: This line needs to be removed before the merge. I am having it to test the servers are working properly simultaneously

**Comment by SidharthBansal** [comment]
Please ignore the codeclimate issues. I don't think they are important.

**Comment by SidharthBansal** [comment]
I have resolved the ones which can be fixed. @jywarren please review this pr.

**Comment by SidharthBansal** [comment]
Three PRs of SWB, Plots2, MK are independent. The login functionality of all 3 apps will not be compromised if any single app breaks. Hurray!!!

**Comment by SidharthBansal** [comment]
I shared the screenshots in the plots2 repo for MK.
Frontend for the SWB is left so there is no frontend changes for SWB yet.
Thanks

**Comment by SidharthBansal** [comment]
The code is similar, not exactly same. SWB and MK both back end are working
properly.

**Comment by SidharthBansal** [comment]
image

We can complete the remaining front end in form of ftos @jywarren else we have completed backend and testing fully. Thanks

**Comment by SidharthBansal** [comment]
The following sequence of events are now present:
1. User enters username at MK then click on login botton
1.1  Public Lab is not logged in. User is redirected to Public Lab /login. User logs into Public Lab successfully through login form. 
1.1.1 If the username entered at MK differs with username entered at Public Lab then the user is shown a flash message
Steps involved in the user interface: 
image
then 
image

then
image

1.1.2 If username entered at MK matches with username entered at Public Lab then user is redirected to MK and is logged in successfully at MK
Steps involved in user interface: 
image
then 
image
then
image

1.2 Public Lab is not logged in. User is redirected to Public Lab /login. User logs into Public Lab successfully through any provider(fb/twitter/github/facebook). 
1.2.1 Provider is logged in already
1.2.1.1 If the username entered at MK differs with  Public Lab then the user is shown a flash message
Steps involved in the user interface: 
image
then user clicks on the provider at 
image

then
image

1.2.1.2 If the username entered at MK matches with Public Lab then the user is redirected to MK and is logged in successfully at MK
Steps involved in user interface: 
image

then user clicks on the provider at 
image

then
image

1.2.2 Provider is not logged in already
1.2.2.1 If the username entered at MK differs with  Public Lab then the user is shown a flash message
Steps involved in the user interface: 
image

then clicks on github provider, lets say
image
then image
then
image


1.2.2.2 If the username entered at MK matches with Public Lab then the user is redirected to MK and is logged in successfully at MK
Steps involved in user interface: 
image
then clicks on github provider, lets say
image
then 

image
then
image
2. clicks on login. Public Lab is logged in.  
2.1 If username entered at MK differs with username entered at Public Lab then the user is shown flash message
Steps involved in user interface: 
image
then
image

2.2 If the username entered at MK matches with username entered at Public Lab then the user is redirected to MK and is logged in successfully at MK
Steps involved in user interface: 
image
then 
image


3. clicks on provider icon[facebook, twitter, google, or github] after filling the username then
3.1 provider is logged in:
3.1.1 Pl.org is logged in:
3.1.1.1 User gets logged into the MK successfully if username matches.
Steps involved in user interface: 
image
then 
image
3.1.1.2 User is not logged into MK as username mismatch
Steps involved in user interface: 
image
then
image


3.1.2 Pl.org is not logged in: 

3.1.2.1 user gets logged into Pl and MK both successfully if username match 
Steps involved in the user interface: 
image
then
image
3.1.2.2 user is shown error message as username mismatch
Steps involved in user interface: 
image
then
image

3.2 provider is not logged in:
3.2.1  Pl.org is logged in: 
3.2.1.1 User gets logged into MK if username matches.
Steps involved in user interface: 
image
then
image

3.2.1.2 User is shown flash msg for username mismatch
Steps involved in user interface: 
image
then
image


3.2.2 Pl.org is not logged in: User authenticates at the provider. 
3.2.2.1 If username matches both at MK and plots2. After successful authentication of the provider, the user gets logged into PL and MK both but redirected to MK.
Steps involved in the user interface: 
image
then
image

then
image
3.2.2.2 if username does not matches at plots2 with the username entered at MK then flash message
image
then
image
then
image

**Comment by SidharthBansal** [comment]
18 flows are listed above. Provider means applicable to all providers. So 18* 4 =72 flows for MK.
Similarly, all 72 flows are working for SWB
Hence all 144 flows are working properly. It was really tough to execute each of them. They are really large in number. But at last all are working properly. Intermediate gateway page is removed. MK and SWB is whitelisted now. Tests are written in the plots2 repository and merged. Frontend main part is made. Rest can be done in ftos. So I am happy!!! :smile: :1st_place_medal:

**Comment by jywarren** [comment]
This is tremendous. I'm going to merge this now!

**Comment by jywarren** [comment]
Great work!!!

I had one thought as follow-up. If people press an OAuth button on MapKnitter, do they really need to input their username? Couldn't we just either a) prompt them for it when they log in on PL.org, or b) use whatever they're already logged into PL.org with as the username?

This isn't super critical but could save one step. But maybe I haven't thought of some shortcoming of this plan...?

**Comment by SidharthBansal** [comment]
Yeah I thought this a lot but openid requires username to get the user
logged in. Without the username the person can't get logged in.
I constantly trying the behaviour you expect from two weeks but could not
arrive at potential solution to it. Will try it again for sure

**Comment by SidharthBansal** [comment]
image
It says
if the user exists with some identity_url
....... the user with some identity_url is logged in
else if it does not exist 
........ create that user with the identity_url.
.........login the user with identity_url

Identity_url is the url which we supply to the plots2 and it contains the username. In case we will not supply the username then it will log same user in for all times due to the database query written above.

I am facing this problem. I got logged in same account X even if I log PL with X, Y or Z.

**Comment by SidharthBansal** [comment]
Finally, my clever insights have awakened. I passed username 'x' always but created the user and logged the user in with the username I received from plots2.
Hurray, we are done with the functionality you want and I also thought of initially but was not able to do that time due to bug.
Hurray!!! @jywarren please see https://github.com/publiclab/mapknitter/pull/575 and https://github.com/publiclab/plots2/pull/5640
thanks

**Comment by jywarren** [comment]
oh whoa, cool!

**Pull request merged**